src/roman_datamodels/datamodels.py

```diff
@@ -322,6 +326,174 @@ class GuidewindowModel(DataModel):
+                self._models.extend(init)
+            else:
+                raise TypeError(
+                        "Input must be a list of strings (full path to ASDF files) or Roman datamodels.")
+        else:
+            raise TypeError(
+                        "Input must be an ASN filepath or list of either strings (full path to ASDF files) or Roman datamodels.")
+
+    def __len__(self):
+        return len(self._models)
+
+    def __getitem__(self, index):
+        m = self._models[index]
+        if not isinstance(m, DataModel) and self._return_open:
+            m = open(m, memmap=self._memmap)
+        return m
+    
+    def __setitem__(self, index, model):
+        self._models[index] = model
+
+    def __delitem__(self, index):
+        del self._models[index]
+
+    def __iter__(self):
+        for model in self._models:
+            if not isinstance(model, DataModel) and self._return_open:
+                model = open(model, memmap=self._memmap)
+            yield model
+
+    def insert(self, index, model):
+        self._models.insert(index, model)
+
+    def append(self, model):
+        self._models.append(model)
+
+    def extend(self, model):
+        self._models.extend(model)
+
+    def pop(self, index=-1):
+        self._models.pop(index)
+
+    def copy(self, memo=None):
+        """
+        Returns a deep copy of the models in this model container.
+        """
+        result = self.__class__(init=None,
+                                pass_invalid_values=self._pass_invalid_values,
+                                strict_validation=self._strict_validation)
+        instance = copy.deepcopy(self._instance, memo=memo)
+        result._asdf = asdf.AsdfFile(instance)
+        result._instance = instance
+        result._iscopy = self._iscopy
+        result._schema = self._schema
+        result._ctx = result
+        for m in self._models:
+            if isinstance(m, DataModel):
+                result.append(m.copy())
+            else:
+                result.append(m)
+        return result
+    
+    @property
+    def models_grouped(self):
+        """
+        Returns a list of a list of datamodels grouped by exposure.
+        Assign an ID grouping by exposure.
+
+        Data from different detectors of the same exposure will have the
+        same group id, which allows grouping by exposure.  The following
+        metadata is used for grouping:
+
+        meta.observation.program
+        meta.observation.observation
+        meta.observation.visit
+        meta.observation.visit_file_group
+        meta.observation.visit_file_sequence
+        meta.observation.visit_file_activity
+        meta.observation.exposure
+        """
+        unique_exposure_parameters = [
+            'program',
+            'observation',
+            'visit',
+            'visit_file_group',
+            'visit_file_sequence',
+            'visit_file_activity',
+            'exposure'
+        ]
```

Comment:
Just because I'm not familiar with these keywords... do we really need all of these to get a unique exposure?  I might have thought that "exposure" alone indicated a unique exposure, or that there were some other exposure ID.

Reply:
Unfortunately exposure does not lead to a unique exposure. The "best" discussion of this is at https://innerspace.stsci.edu/display/RI/Roman+APT

Reply:
Right, I am permanently confused here.  The associations are generated at APT time when we haven't actually done the observations yet.  Is it the case that _after_ an observation is made, it has some unique exposure ID associated with it?  e.g., this was exposure number 123456789 taken by Roman?  For things like this where we have the actual L1 images it feels like we should have access to that kind of ID?